File: dot_local/share/herdr-plugins/workspace-mru/workspace-mru.test.mjs
import assert from "node:assert/strict";
import test from "node:test";

import {
	recordWorkspaceFocus,
	removeWorkspace,
	selectNextWorkspace,
} from "./workspace-mru.mjs";

function workspaces(focusedId, ids = ["A", "B", "C"]) {
	return ids.map((workspaceId) => ({
		workspace_id: workspaceId,
		focused: workspaceId === focusedId,
	}));
}

test("records workspaces from most to least recently used", () => {
	let state = {};
	state = recordWorkspaceFocus(state, "B");
	state = recordWorkspaceFocus(state, "C");
	state = recordWorkspaceFocus(state, "A");
	assert.deepEqual(state.mru, ["A", "C", "B"]);
});

test("uses MRU first and sidebar order for continued presses", () => {
	let state = {};
	for (const id of ["B", "C", "A"]) state = recordWorkspaceFocus(state, id);

	let selection = selectNextWorkspace(state, workspaces("A"), 1000);
	assert.equal(selection.target, "C");
	state = recordWorkspaceFocus(selection.state, "C");

	selection = selectNextWorkspace(state, workspaces("C"), 1250);
	assert.equal(selection.target, "A");
	state = recordWorkspaceFocus(selection.state, "A");

	selection = selectNextWorkspace(state, workspaces("A"), 1499);
	assert.equal(selection.target, "B");
});

test("starts a new MRU cycle when 500ms elapse", () => {
	let state = {};
	for (const id of ["B", "C", "A"]) state = recordWorkspaceFocus(state, id);
	let selection = selectNextWorkspace(state, workspaces("A"), 1000);
	state = recordWorkspaceFocus(selection.state, "C");

	selection = selectNextWorkspace(state, workspaces("C"), 1500);
	assert.equal(selection.target, "A");
	assert.deepEqual(selection.state.mru, ["C", "A", "B"]);
});

test("starts a fresh cycle after a workspace is focused another way", () => {
	let state = {};
	for (const id of ["B", "C", "A"]) state = recordWorkspaceFocus(state, id);
	let selection = selectNextWorkspace(state, workspaces("A"), 1000);
	state = recordWorkspaceFocus(selection.state, "C");
	state = recordWorkspaceFocus(state, "B");

	selection = selectNextWorkspace(state, workspaces("B"), 1100);
	assert.equal(selection.target, "A");
});

test("removes closed workspaces from history and an active cycle", () => {
	let state = {};
	for (const id of ["C", "B", "A"]) state = recordWorkspaceFocus(state, id);
	state = selectNextWorkspace(state, workspaces("A"), 1000).state;
	state = removeWorkspace(state, "C");
	assert.deepEqual(state.mru, ["A", "B"]);
	assert.deepEqual(state.cycle.order, ["A", "B"]);
	assert.deepEqual(state.pending, ["B"]);
});

File: dot_local/share/herdr-plugins/workspace-mru/workspace_mru_test.go
package main

import (
	"reflect"
	"testing"
)

func testWorkspaces(focusedID string, ids ...string) []HerdrWorkspaceRow {
	if len(ids) == 0 {
		ids = []string{"A", "B", "C"}
	}
	rows := make([]HerdrWorkspaceRow, len(ids))
	for index, id := range ids {
		rows[index] = HerdrWorkspaceRow{WorkspaceID: id, Focused: id == focusedID}
	}
	return rows
}

func TestRecordWorkspaceFocus(t *testing.T) {
	var state WorkspaceMRUState
	state = RecordWorkspaceFocus(state, "B")
	state = RecordWorkspaceFocus(state, "C")
	state = RecordWorkspaceFocus(state, "A")
	if !reflect.DeepEqual(state.MRU, []string{"A", "C", "B"}) {
		t.Fatalf("got %#v", state.MRU)
	}
}

func TestSelectNextWorkspaceMRUThenSidebar(t *testing.T) {
	var state WorkspaceMRUState
	for _, id := range []string{"B", "C", "A"} {
		state = RecordWorkspaceFocus(state, id)
	}

	state, target := SelectNextWorkspace(state, testWorkspaces("A"), 1000)
	if target != "C" {
		t.Fatalf("got %q", target)
	}
	state = RecordWorkspaceFocus(state, "C")

	state, target = SelectNextWorkspace(state, testWorkspaces("C"), 1250)
	if target != "A" {
		t.Fatalf("got %q", target)
	}
	state = RecordWorkspaceFocus(state, "A")

	_, target = SelectNextWorkspace(state, testWorkspaces("A"), 1499)
	if target != "B" {
		t.Fatalf("got %q", target)
	}
}

func TestSelectNextWorkspaceStartsFreshAfter500ms(t *testing.T) {
	var state WorkspaceMRUState
	for _, id := range []string{"B", "C", "A"} {
		state = RecordWorkspaceFocus(state, id)
	}
	state, _ = SelectNextWorkspace(state, testWorkspaces("A"), 1000)
	state = RecordWorkspaceFocus(state, "C")

	state, target := SelectNextWorkspace(state, testWorkspaces("C"), 1500)
	if target != "A" {
		t.Fatalf("got %q", target)
	}
	if !reflect.DeepEqual(state.MRU, []string{"C", "A", "B"}) {
		t.Fatalf("got %#v", state.MRU)
	}
}

func TestSelectNextWorkspaceStartsFreshAfterOtherFocus(t *testing.T) {
	var state WorkspaceMRUState
	for _, id := range []string{"B", "C", "A"} {
		state = RecordWorkspaceFocus(state, id)
	}
	state, _ = SelectNextWorkspace(state, testWorkspaces("A"), 1000)
	state = RecordWorkspaceFocus(state, "C")
	state = RecordWorkspaceFocus(state, "B")

	_, target := SelectNextWorkspace(state, testWorkspaces("B"), 1100)
	if target != "A" {
		t.Fatalf("got %q", target)
	}
}

func TestRemoveWorkspaceFromHistoryAndCycle(t *testing.T) {
	var state WorkspaceMRUState
	for _, id := range []string{"C", "B", "A"} {
		state = RecordWorkspaceFocus(state, id)
	}
	state, _ = SelectNextWorkspace(state, testWorkspaces("A"), 1000)
	state = RemoveWorkspace(state, "C")
	if !reflect.DeepEqual(state.MRU, []string{"A", "B"}) {
		t.Fatalf("mru %#v", state.MRU)
	}
	if state.Cycle == nil || !reflect.DeepEqual(state.Cycle.Order, []string{"A", "B"}) {
		t.Fatalf("cycle %#v", state.Cycle)
	}
	if !reflect.DeepEqual(state.Pending, []string{"B"}) {
		t.Fatalf("pending %#v", state.Pending)
	}
}
